cpp/src/arrow/scalar.h

```diff
@@ -49,7 +49,8 @@ class Array;
 /// Scalars are useful for passing single value inputs to compute functions,
 /// or for representing individual array elements (with a non-trivial
 /// wrapping cost, though).
-struct ARROW_EXPORT Scalar : public util::EqualityComparable<Scalar> {
+struct ARROW_EXPORT Scalar : public std::enable_shared_from_this<Scalar>,
```

Comment:
IMHO this isn't wonderful. I would rather introduce `virtual std::shared_ptr<DataType> Scalar::Copy()`

Reply:
we could rename the helper function for now and patch up the implementation later

Reply:
done